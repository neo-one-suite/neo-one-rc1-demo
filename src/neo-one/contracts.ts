/* @hash a00f124f9fcefc1169e1794d341102e1 */
// tslint:disable
/* eslint-disable */
import { Client } from '@neo-one/client';

import { TestTokenSmartContract, TestTokenMigrationSmartContract } from './TestToken/types';

import { createTestTokenSmartContract } from './TestToken/contract';

export interface Contracts<TClient extends Client = Client> {
  readonly testToken: TestTokenSmartContract<TClient>;
}
// Refer to the MigrationSmartContract documentation at https://neo-one.io/docs/deployment for more information.
export interface MigrationContracts {
  readonly testToken: TestTokenMigrationSmartContract;
}

export const createContracts = <TClient extends Client>(client: TClient): Contracts<TClient> => ({
  testToken: createTestTokenSmartContract(client),
});
