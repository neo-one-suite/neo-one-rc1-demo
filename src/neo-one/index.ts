/* @hash 8e1b2db3e3af186e5b1c39ae7dd9952a */
// tslint:disable
/* eslint-disable */
export * from './contracts';
export * from './react';
export * from './client';
export * from './TestToken/contract';
export * from './TestToken/types';
export * from './TestToken/manifest';
