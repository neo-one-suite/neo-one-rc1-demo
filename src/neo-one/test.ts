/* @hash ab6704f8b9f2320fffe2284a559cfcf1 */
// tslint:disable
/* eslint-disable */
import { createWithContracts, TestOptions, WithContractsOptions } from '@neo-one/smart-contract-test';
import { Contracts } from './contracts';
import * as path from 'path';

export const withContracts: (
  test: (contracts: Contracts & TestOptions) => Promise<void>,
  options?: WithContractsOptions,
) => Promise<void> = createWithContracts([
  { name: 'TestToken', filePath: path.resolve(__dirname, '../../neo-one/contracts/TestToken.ts') },
]);
