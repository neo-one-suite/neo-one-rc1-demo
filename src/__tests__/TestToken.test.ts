import { privateKeyToAddress } from '@neo-one/client';
import { withContracts } from '../neo-one/test';
import BigNumber from 'bignumber.js';

const TO = {
  PRIVATE_KEY: '536f1e9f0466f6cd5b2ea5374d00f038786daa0f0e892161d6b0cb4d6b154740',
  PUBLIC_KEY: '03463b7a0afc41ff1f6a386190f99bafd1deca48f4026aeac95435731af278cb7d',
};

const ZERO = {
  PRIVATE_KEY: '9c111f04a34b3a07600fe701d308dce6e20c86268c105f21c2f30e9fef7e7968',
  PUBLIC_KEY: '027f73dbc47133b08a4bc0fc04589fc76525baaf3bebe71bdd78053d559c41db70',
};

jest.setTimeout(30000);

describe('Test TestToken methods', () => {
  test('smart contract', async () => {
    await withContracts(async ({ testToken, client, networkName, masterAccountID }) => {
      const transferValue = new BigNumber('10');
      const issueValue = new BigNumber('100');

      expect(testToken).toBeDefined();

      const [nameResult, decimalsResult, symbolResult, wallet0] = await Promise.all([
        testToken.name(),
        testToken.decimals(),
        testToken.symbol(),
        client.providers.memory.keystore.addUserAccount({
          network: networkName,
          name: 'wallet0',
          privateKey: TO.PRIVATE_KEY,
        }),
      ]);
      expect(nameResult).toEqual('TestToken');
      expect(decimalsResult.toString()).toEqual('8');
      expect(symbolResult).toEqual('TT');

      const account0 = wallet0.userAccount.id;

      const [issueBalance, issueTotalSupply, transferResult] = await Promise.all([
        testToken.balanceOf(masterAccountID.address),
        testToken.totalSupply(),
        testToken.transfer(masterAccountID.address, account0.address, transferValue, { from: masterAccountID }),
      ]);
      expect(issueBalance.toString()).toEqual(issueValue.toString());
      expect(issueTotalSupply.toString()).toEqual(issueValue.toString());

      const transferReceipt = await transferResult.confirmed({
        timeoutMS: 2500,
      });

      if (transferReceipt.result.state !== 'HALT') {
        throw new Error(transferReceipt.result.message);
      }

      expect(transferReceipt.result.gasConsumed.toString()).toMatchSnapshot('transfer consume');
      expect(transferReceipt.events).toHaveLength(1);
      const event = transferReceipt.events[0];
      expect(event.name).toEqual('Transfer');
      expect(event.parameters.from).toEqual(masterAccountID.address);
      expect(event.parameters.to).toEqual(account0.address);
      if (event.parameters.amount === undefined) {
        expect(event.parameters.amount).toBeTruthy();
        throw new Error('For TS');
      }

      expect(event.parameters.amount.toString()).toEqual(transferValue.toString());

      const [
        transferMasterBalance,
        transferAccountBalance,
        transferTotalSupply,
        transferZeroBalance,
      ] = await Promise.all([
        testToken.balanceOf(masterAccountID.address),
        testToken.balanceOf(account0.address),
        testToken.totalSupply(),
        testToken.balanceOf(privateKeyToAddress(ZERO.PRIVATE_KEY)),
      ]);

      const remainingValue = issueValue.minus(transferValue);
      expect(transferMasterBalance.toString()).toEqual(remainingValue.toString());
      expect(transferAccountBalance.toString()).toEqual(transferValue.toString());
      expect(transferTotalSupply.toString()).toEqual(issueValue.toString());
      expect(transferZeroBalance.toString()).toEqual('0');

      const readClient = client.read(networkName);
      const contract = await readClient.getContract(testToken.definition.networks[networkName].address);

      expect(contract.id).toEqual(1);
      expect(contract.updateCounter).toEqual(0);
      expect(contract.nef.compiler).toEqual('neo-one 3.0.0-rc1');
      expect(contract.manifest.abi.events).toMatchSnapshot('contract events');
      expect(contract.manifest.abi.methods).toMatchSnapshot('contract methods');
      expect(contract.manifest.name).toEqual('TestToken');
      expect(contract.manifest.supportedStandards).toContain('NEP-17');
      expect(contract.manifest.trusts).toEqual('*');
      expect(contract.manifest.extra).toEqual({});
      expect(contract.manifest.groups).toEqual([]);
      expect(contract.manifest.permissions).toEqual([]);
    });
  });
});
