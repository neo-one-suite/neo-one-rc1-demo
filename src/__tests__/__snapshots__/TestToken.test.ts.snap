// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TestToken methods smart contract: contract events 1`] = `
Array [
  Object {
    "name": "Transfer",
    "parameters": Array [
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "to",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
    ],
  },
  Object {
    "name": "approveSendTransfer",
    "parameters": Array [
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "to",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
    ],
  },
  Object {
    "name": "revokeSendTransfer",
    "parameters": Array [
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "to",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
    ],
  },
]
`;

exports[`Test TestToken methods smart contract: contract methods 1`] = `
Array [
  Object {
    "name": "owner",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
    ],
    "returnType": "Hash160",
    "safe": true,
  },
  Object {
    "name": "name",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
    ],
    "returnType": "String",
    "safe": true,
  },
  Object {
    "name": "decimals",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
    ],
    "returnType": "Integer",
    "safe": true,
  },
  Object {
    "name": "symbol",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
    ],
    "returnType": "String",
    "safe": true,
  },
  Object {
    "name": "totalSupply",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
    ],
    "returnType": "Integer",
    "safe": true,
  },
  Object {
    "name": "balanceOf",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "address",
        "type": "Hash160",
      },
    ],
    "returnType": "Integer",
    "safe": true,
  },
  Object {
    "name": "approvedTransfer",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "to",
        "type": "Hash160",
      },
    ],
    "returnType": "Integer",
    "safe": true,
  },
  Object {
    "name": "transfer",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "to",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
    ],
    "returnType": "Boolean",
    "safe": false,
  },
  Object {
    "name": "approveSendTransfer",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "to",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
    ],
    "returnType": "Boolean",
    "safe": false,
  },
  Object {
    "name": "approveReceiveTransfer",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
      Object {
        "name": "asset",
        "type": "Hash160",
      },
    ],
    "returnType": "Boolean",
    "safe": false,
  },
  Object {
    "name": "onNEP17Payment",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
      Object {
        "name": "data",
        "type": "Buffer",
      },
    ],
    "returnType": "Void",
    "safe": false,
  },
  Object {
    "name": "revokeSendTransfer",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "to",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
    ],
    "returnType": "Boolean",
    "safe": false,
  },
  Object {
    "name": "onRevokeSendTransfer",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "from",
        "type": "Hash160",
      },
      Object {
        "name": "amount",
        "type": "Integer",
      },
      Object {
        "name": "asset",
        "type": "Hash160",
      },
    ],
    "returnType": "Void",
    "safe": false,
  },
  Object {
    "name": "deploy",
    "offset": 0,
    "parameters": Array [
      Object {
        "name": "__neoonemethodstring",
        "type": "String",
      },
      Object {
        "name": "owner",
        "type": "Hash160",
      },
    ],
    "returnType": "Boolean",
    "safe": false,
  },
]
`;

exports[`Test TestToken methods smart contract: transfer consume 1`] = `"124212640"`;
